File: ZYH.WebControl_V/Js/TextBox.js
function Class_TextBox(ClientID, CallbackOptions, Settings) {
    var Self = this;
    Self.Control = $('#' + ClientID);
    Self.Settings = Settings;

    var TempTextBoxBackColor;
    var TextBoxLastKeyUp_e_BeforeCall;
    var TextBoxLastKeyUp_Time;

    Self.Serialize = function (e) {
        var xml = '<TextBox>';
        var Text = Self.Control.val(); if (Text != '') xml += '<Text>' + htmlEncode(Text) + '</Text>';
        if (e.keyCode && e.keyCode != 0) xml += '<KeyCode>' + e.keyCode + '</KeyCode>';
        if (e.ctrlKey) xml += '<Ctrl>' + e.ctrlKey + '</Ctrl>';
        if (e.altKey) xml += '<Alt>' + e.ctrlKey + '</Alt>';
        if (e.shiftKey) xml += '<Shift>' + e.shiftKey + '</Shift>';
        if (e.type != 'blur') var CP = Self.GetCursorPosition(); if (CP) xml += '<CP>' + CP + '</CP>';
        xml += '</TextBox>';
        return xml;
    }

    Self.Init = function () {
        Self.Control.bind('focus', Self.GetFocus);
        Self.Control.bind('blur', Self.LostFocus);
        Self.Control.bind('keydown', Self.KeyDown);
        Self.Control.bind('keyup', Self.KeyUp);
        Self.Control.bind('paste', function (e) { setTimeout(function () { Self.KeyUp(e); }, 5); });
    }

    Self.AfterCallProcessing = function (id, e_AfterCall) {
        e_AfterCall.CommandArg = Self.Settings.CommandArg;
        var DOM = XMLtoDOM(e_AfterCall.ControlReserved);
        e_AfterCall.Text = $('Text', DOM).text();
        var KeyCode = $('KeyCode', DOM).text(); if (KeyCode == '') e_AfterCall.KeyCode = 0; else e_AfterCall.KeyCode = parseInt(KeyCode);
        var Ctrl = $('Ctrl', DOM).text(); if (Ctrl == '') e_AfterCall.Ctrl = false; else e_AfterCall.Ctrl = true;
        var Alt = $('Alt', DOM).text(); if (Alt == '') e_AfterCall.Alt = false; else e_AfterCall.Alt = true;
        var Shift = $('Shift', DOM).text(); if (Shift == '') e_AfterCall.Shift = false; else e_AfterCall.Shift = true;
        var CP = $('CP', DOM).text(); if (CP == '') e_AfterCall.CursorPosition = 0; else e_AfterCall.CursorPosition = parseInt(CP);
        var SCPT = $('SCPT', DOM).text(); if (SCPT == '') e_AfterCall.SetCursorPositionTo = -1; else e_AfterCall.SetCursorPositionTo = parseInt(SCPT);
        switch (e_AfterCall.CallbackOptions.AfterCallAction) {
            case 0://'NA':
                break;
            case 11://'RefreshToSynchro':
                if (Self.Control.val() != e_AfterCall.Text) {
                    Self.Control.val(e_AfterCall.Text);
                    Self.SetCursorPosition(e_AfterCall.CursorPosition);
                }
                break;
            case 1://'RecallToSynchro':
                if (Self.Control.val() != e_AfterCall.Text) { Self.OnKeyUpDeply(); }
                break;
        }
        if (e_AfterCall.SetCursorPositionTo != -1) Self.SetCursorPosition(e_AfterCall.SetCursorPositionTo); //not for LostFocus
    }

    Self.GetFocus = function (e) {
        Self.GFCC();
        //Call Server
        var FunBeforeCall = null;
        var FunAfterCall = null;
        var e_BeforeCall = { 'CancelCall': !Self.Settings.ServerEvents_GetFocus, 'ArgFromClient': '', 'CommandArg': Self.Settings.CommandArg, 'Instance': Self, 'EventTriggerHierarchyID': Self.Settings.EventTriggerHierarchyID, 'CallbackOptions': CallbackOptions };
        e_BeforeCall.ControlReserved = Self.Serialize(e);
        e_BeforeCall.Text = Self.Control.val();
        if (Self.Settings.ClientEvent_BeforeCallback_GetFocus != '') { FunBeforeCall = eval(Self.Settings.ClientEvent_BeforeCallback_GetFocus); }
        if (Self.Settings.ClientEvent_AfterCallback_GetFocus != '') { FunAfterCall = eval(Self.Settings.ClientEvent_AfterCallback_GetFocus); }
        eval(Self.Settings.CallbackManagerClientInstance).CallServer(Self.Settings.UniqueID, ClientID, '8', FunBeforeCall, e_BeforeCall, Self.AfterCallProcessing, FunAfterCall);

    }

    Self.GFCC = function () {
        TempTextBoxBackColor = Self.Control.css('background-color');
        Self.Control.css('background-color', Self.Settings.OnFocusBackColor);
    }

    Self.LostFocus = function (e) {
        Self.LFCC();
        //Call Server
        var FunBeforeCall = null;
        var FunAfterCall = null;
        var e_BeforeCall = { 'CancelCall': !Self.Settings.ServerEvents_LostFocus, 'ArgFromClient': '', 'CommandArg': Self.Settings.CommandArg, 'Instance': Self, 'EventTriggerHierarchyID': Self.Settings.EventTriggerHierarchyID, 'CallbackOptions': CallbackOptions };
        e_BeforeCall.ControlReserved = Self.Serialize(e);
        e_BeforeCall.Text = Self.Control.val();
        if (Self.Settings.ClientEvent_BeforeCallback_LostFocus != '') { FunBeforeCall = eval(Self.Settings.ClientEvent_BeforeCallback_LostFocus); }
        if (Self.Settings.ClientEvent_AfterCallback_LostFocus != '') { FunAfterCall = eval(Self.Settings.ClientEvent_AfterCallback_LostFocus); }
        eval(Self.Settings.CallbackManagerClientInstance).CallServer(Self.Settings.UniqueID, ClientID, '9', FunBeforeCall, e_BeforeCall, Self.AfterCallProcessing, FunAfterCall);

    }

    Self.LFCC = function () {
        Self.Control.css('background-color', TempTextBoxBackColor);
    }

    Self.KeyDown = function (e) {
        //Call Server
        var FunBeforeCall = null;
        var FunAfterCall = null;
        var e_BeforeCall = { 'CancelCall': !Self.Settings.ServerEvents_KeyDown, 'ArgFromClient': '', 'CommandArg': Self.Settings.CommandArg, 'Instance': Self, 'EventTriggerHierarchyID': Self.Settings.EventTriggerHierarchyID, 'CallbackOptions': CallbackOptions };
        e_BeforeCall.ControlReserved = Self.Serialize(e);
        e_BeforeCall.Text = Self.Control.val();
        e_BeforeCall.KeyCode = e.keyCode;
        e_BeforeCall.Ctrl = e.ctrlKey;
        e_BeforeCall.Alt = e.altKey;
        e_BeforeCall.Shift = e.shiftKey;
        if (Self.Settings.ClientEvent_BeforeCallback_KeyDown != '') { FunBeforeCall = eval(Self.Settings.ClientEvent_BeforeCallback_KeyDown); }
        if (Self.Settings.ClientEvent_AfterCallback_KeyDown != '') { FunAfterCall = eval(Self.Settings.ClientEvent_AfterCallback_KeyDown); }
        eval(Self.Settings.CallbackManagerClientInstance).CallServer(Self.Settings.UniqueID, ClientID, '6', FunBeforeCall, e_BeforeCall, Self.AfterCallProcessing, FunAfterCall);
    }

    Self.KeyUp = function (e) {
        TextBoxLastKeyUp_e_BeforeCall = { 'CancelCall': !Self.Settings.ServerEvents_KeyUp, 'ArgFromClient': '', 'CommandArg': Self.Settings.CommandArg, 'Instance': Self, 'EventTriggerHierarchyID': Self.Settings.EventTriggerHierarchyID, 'CallbackOptions': CallbackOptions };
        TextBoxLastKeyUp_e_BeforeCall.ControlReserved = Self.Serialize(e);
        TextBoxLastKeyUp_e_BeforeCall.Text = Self.Control.val();
        TextBoxLastKeyUp_e_BeforeCall.KeyCode = e.keyCode;
        TextBoxLastKeyUp_e_BeforeCall.Ctrl = e.ctrlKey;
        TextBoxLastKeyUp_e_BeforeCall.Alt = e.altKey;
        TextBoxLastKeyUp_e_BeforeCall.Shift = e.shiftKey;

        TextBoxLastKeyUp_Time = new Date();
        setTimeout(Self.OnKeyUpDeply, Self.Settings.KeyUpEventDelay + 50);
    }

    Self.OnKeyUpDeply = function () {
        if (new Date().getTime() - TextBoxLastKeyUp_Time.getTime() < Self.Settings.KeyUpEventDelay) { return; }

        var FunBeforeCall = null;
        var FunAfterCall = null;
        if (Self.Settings.ClientEvent_BeforeCallback_KeyUp != '') { FunBeforeCall = eval(Self.Settings.ClientEvent_BeforeCallback_KeyUp); }
        if (Self.Settings.ClientEvent_AfterCallback_KeyUp != '') { FunAfterCall = eval(Self.Settings.ClientEvent_AfterCallback_KeyUp); }
        eval(Self.Settings.CallbackManagerClientInstance).CallServer(Self.Settings.UniqueID, ClientID, '7', FunBeforeCall, TextBoxLastKeyUp_e_BeforeCall, Self.AfterCallProcessing, FunAfterCall);
    }

    Self.GetText = function () {
        return Self.Control.val();
    }

    Self.SetText = function (txt) {
        Self.Control.val(txt);
    }

    Self.GetCursorPosition = function () {
        var ctrl = Self.Control[0];
        var CaretPos = 0;
        // IE 
        if (document.selection) {
            ctrl.focus();
            var Sel = document.selection.createRange();
            Sel.moveStart('character', -ctrl.value.length);
            CaretPos = Sel.text.length;
        }
            // Firefox 
        else if (ctrl.selectionStart || ctrl.selectionStart == '0')
            CaretPos = ctrl.selectionStart;
        return (CaretPos);
    }

    Self.SetCursorPosition = function (pos) {
        var ctrl = Self.Control[0];
        if (ctrl.setSelectionRange) {
            ctrl.focus();
            ctrl.setSelectionRange(pos, pos);
        } else if (ctrl.createTextRange) {
            var range = ctrl.createTextRange();
            range.collapse(true);
            range.moveEnd('character', pos);
            range.moveStart('character', pos);
            range.select();
        }
    }

}

File: ZYH.WebControl_V/TextBox.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ComponentModel;

namespace ZYH.WebControl_V
{
    [DefaultEvent("KeyUp"), System.Drawing.ToolboxBitmap(typeof(TextBox), "Image.ToolBox.TextBox.bmp")]
    public class TextBox : CallbackControl_Base
    {
        public System.Web.UI.WebControls.TextBox _TextBox = new System.Web.UI.WebControls.TextBox();
        private TextBoxData _Data = new TextBoxData();
        private TextBoxSettings _Settings = new TextBoxSettings();

        #region Event
        public delegate void KeyDownHandler(object sender, TextBoxEventArg e);
        public event KeyDownHandler KeyDown;
        public delegate void KeyUpHandler(object sender, TextBoxEventArg e);
        /// <summary>
        /// This event is fired by either keyup or paste. e.KeyCode=0 when fired by paste.
        /// </summary>
        public event KeyUpHandler KeyUp;
        public delegate void GetFocusHandler(object sender, TextBoxEventArg e);
        public event GetFocusHandler GetFocus;
        public delegate void LostFocusHandler(object sender, TextBoxEventArg e);
        public event LostFocusHandler LostFocus;

        public override void OnRaiseCallbackEvent(CallbackParameters CallbackParameters, CallbackResult CallbackResult)
        {
            TextBoxEventArg e = new TextBoxEventArg();
            e.CallbackManager = this.CallbackManager;
            e.ArgFromClient = CallbackParameters.ArgFromClient;
            e.Command = CallbackParameters.Command;
            e.CommandArg = this.CommandArg;
            e.CallbackOptions = this._CallbackOptions;
            e.Text = this.Text;
            e.Alt = _Data.Alt;
            e.Ctrl = _Data.Ctrl;
            e.Shift = _Data.Shift;
            e.KeyCode = _Data.KeyCode;
            e.CursorPosition = _Data.CursorPosition;
            switch (e.Command)
            {
                case CallbackCommands.TextBox_KeyDown:
                    OnKeyDown(e);
                    break;
                case CallbackCommands.TextBox_KeyUp:
                    OnKeyUp(e);
                    break;
                case CallbackCommands.TextBox_GetFocus:
                    OnGetFocus(e);
                    break;
                case CallbackCommands.TextBox_LostFocus:
                    OnLostFocus(e);
                    break;
            }
            CallbackResult.ArgToClient = e.ArgToClient;
            CallbackResult.CallbackOptions = e.CallbackOptions;
            _Data.Alt = e.Alt;
            _Data.Ctrl = e.Ctrl;
            _Data.Shift = e.Shift;
            _Data.KeyCode = e.KeyCode;
            _Data.Text = e.Text;
            _Data.CursorPosition = e.CursorPosition;
            CallbackResult.ControlReserved = _Data.ToXML();
        }

        public virtual void OnKeyDown(TextBoxEventArg e) { if (KeyDown != null) KeyDown(this, e); }

        public virtual void OnKeyUp(TextBoxEventArg e) { if (KeyUp != null) KeyUp(this, e); }

        public virtual void OnGetFocus(TextBoxEventArg e) { if (GetFocus != null) GetFocus(this, e); }

        public virtual void OnLostFocus(TextBoxEventArg e) { if (LostFocus != null) LostFocus(this, e); }
        #endregion

        #region Property
        [DefaultValue(typeof(System.Web.UI.WebControls.TextBoxMode), "SingleLine")]
        public System.Web.UI.WebControls.TextBoxMode TextMode
        {
            get { return _TextBox.TextMode; }
            set { _TextBox.TextMode = value; }
        }

        [DefaultValue("")]
        public string Text
        {
            get { return _TextBox.Text; }
            set { _TextBox.Text = value; }
        }

        [DefaultValue(false)]
        public bool ReadOnly
        {
            get { return _TextBox.ReadOnly; }
            set { _TextBox.ReadOnly = value; }
        }

        [DefaultValue(true)]
        public bool Wrap
        {
            get { return _TextBox.Wrap; }
            set { _TextBox.Wrap = value; }
        }

        [DefaultValue(0)]
        public int Columns
        {
            get { return _TextBox.Columns; }
            set { _TextBox.Columns = value; }
        }

        [DefaultValue(0)]
        public int Rows
        {
            get { return _TextBox.Rows; }
            set { _TextBox.Rows = value; }
        }

        [DefaultValue(0)]
        public int MaxLength
        {
            get { return _TextBox.MaxLength; }
            set { _TextBox.MaxLength = value; }
        }

        [DefaultValue(0)]
        public override short TabIndex
        {
            get { return _TextBox.TabIndex; }
            set { _TextBox.TabIndex = value; }
        }

        [DefaultValue(200), Category("Behavior"), Description("if > 0, delay the event until the value of minisecond after last key up.")]
        public int KeyUpEventDelay
        {
            get { return _Settings.KeyUpEventDelay; }
            set { _Settings.KeyUpEventDelay = value; }
        }

        [Category("Appearance"), Description("Get or set alternating back color of text box."), DefaultValue(typeof(System.Drawing.Color), "MintCream"), RefreshProperties(RefreshProperties.All), TypeConverter(typeof(System.Web.UI.WebControls.WebColorConverter))]
        public System.Drawing.Color OnFocusBackColor
        {
            get { return System.Drawing.ColorTranslator.FromHtml(_Settings.OnFocusBackColor); }
            set { _Settings.OnFocusBackColor = System.Drawing.ColorTranslator.ToHtml(value); }
        }

        private System.Drawing.Color _ReadOnlyBackColor = System.Drawing.Color.PapayaWhip;
        [Category("Appearance"), Description("Get or set alternating back color of text box."), DefaultValue(typeof(System.Drawing.Color), "PapayaWhip"), RefreshProperties(RefreshProperties.All), TypeConverter(typeof(System.Web.UI.WebControls.WebColorConverter))]
        public System.Drawing.Color ReadOnlyBackColor
        {
            get { return _ReadOnlyBackColor; }
            set { _ReadOnlyBackColor = value; }
        }

        #region Client Event
        [Category("Behavior"), Description("not applicable."), DefaultValue("not applicable")]
        public new string ClientEvent_BeforeCallback { get { return "not applicable"; } }

        [Category("Behavior"), Description("not applicable."), DefaultValue("not applicable")]
        public new string ClientEvent_AfterCallback { get { return "not applicable"; } }

        [Category("Behavior"), DefaultValue(""), Description("function BeforeCall(SenderID, e_BeforeCall) {} var e_BeforeCall = new Object(); e_BeforeCall.CancelCall. = false; e_BeforeCall.ArgFromClient = ''; e_BeforeCall.ControlReserved = ControlReserved; e_BeforeCall.Text = text; ")]
        public string ClientEvent_BeforeCallback_GetFocus
        {
            get { return _Settings.ClientEvent_BeforeCallback_GetFocus; }
            set { _Settings.ClientEvent_BeforeCallback_GetFocus = value; }
        }

        [Category("Behavior"), DefaultValue(""), Description("function AfterCall(SenderID, e_AfterCall) {} var e_AfterCall = new Object(); e_AfterCall.CancelCall. = false; e_AfterCall.ArgFromClient = ''; e_AfterCall.ControlReserved = ControlReserved; e_AfterCall.Text = text; ")]
        public string ClientEvent_AfterCallback_GetFocus
        {
            get { return _Settings.ClientEvent_AfterCallback_GetFocus; }
            set { _Settings.ClientEvent_AfterCallback_GetFocus = value; }
        }

        [Category("Behavior"), DefaultValue(""), Description("function BeforeCall(SenderID, e_BeforeCall) {} var e_BeforeCall = new Object(); e_BeforeCall.CancelCall. = false; e_BeforeCall.ArgFromClient = ''; e_BeforeCall.ControlReserved = ControlReserved; e_BeforeCall.Text = text; ")]
        public string ClientEvent_BeforeCallback_LostFocus
        {
            get { return _Settings.ClientEvent_BeforeCallback_LostFocus; }
            set { _Settings.ClientEvent_BeforeCallback_LostFocus = value; }
        }

        [Category("Behavior"), DefaultValue(""), Description("function AfterCall(SenderID, e_AfterCall) {} var e_AfterCall = new Object(); e_AfterCall.CancelCall. = false; e_AfterCall.ArgFromClient = ''; e_AfterCall.ControlReserved = ControlReserved; e_AfterCall.Text = text; ")]
        public string ClientEvent_AfterCallback_LostFocus
        {
            get { return _Settings.ClientEvent_AfterCallback_LostFocus; }
            set { _Settings.ClientEvent_AfterCallback_LostFocus = value; }
        }

        [Category("Behavior"), DefaultValue(""), Description("function BeforeCall(SenderID, e_BeforeCall) {} var e_BeforeCall = new Object(); e_BeforeCall.CancelCall. = false; e_BeforeCall.ArgFromClient = ''; e_BeforeCall.ControlReserved = ControlReserved; e_BeforeCall.Text = text; ")]
        public string ClientEvent_BeforeCallback_KeyDown
        {
            get { return _Settings.ClientEvent_BeforeCallback_KeyDown; }
            set { _Settings.ClientEvent_BeforeCallback_KeyDown = value; }
        }

        [Category("Behavior"), DefaultValue(""), Description("function AfterCall(SenderID, e_AfterCall) {} var e_AfterCall = new Object(); e_AfterCall.CancelCall. = false; e_AfterCall.ArgFromClient = ''; e_AfterCall.ControlReserved = ControlReserved; e_AfterCall.Text = text; ")]
        public string ClientEvent_AfterCallback_KeyDown
        {
            get { return _Settings.ClientEvent_AfterCallback_KeyDown; }
            set { _Settings.ClientEvent_AfterCallback_KeyDown = value; }
        }

        [Category("Behavior"), DefaultValue(""), Description("This event is fired by either keyup or paste. e.KeyCode=0 when fired by paste. function BeforeCall(SenderID, e_BeforeCall) {} var e_BeforeCall = new Object(); e_BeforeCall.CancelCall. = false; e_BeforeCall.ArgFromClient = ''; e_BeforeCall.ControlReserved = ControlReserved; e_BeforeCall.Text = text; ")]
        public string ClientEvent_BeforeCallback_KeyUp
        {
            get { return _Settings.ClientEvent_BeforeCallback_KeyUp; }
            set { _Settings.ClientEvent_BeforeCallback_KeyUp = value; }
        }

        [Category("Behavior"), DefaultValue(""), Description("This event is fired by either keyup or paste. e.KeyCode=0 when fired by paste. function AfterCall(SenderID, e_AfterCall) {} var e_AfterCall = new Object(); e_AfterCall.CancelCall. = false; e_AfterCall.ArgFromClient = ''; e_AfterCall.ControlReserved = ControlReserved; e_AfterCall.Text = text; ")]
        public string ClientEvent_AfterCallback_KeyUp
        {
            get { return _Settings.ClientEvent_AfterCallback_KeyUp; }
            set { _Settings.ClientEvent_AfterCallback_KeyUp = value; }
        }


        #endregion

        #endregion

        private void InitPostData()
        {
            string st = Page.Request.Params[UniqueID];
            if (st != null) { Text = st; }
        }

        protected override void OnInit(EventArgs e)
        {
            base.OnInit(e);

            if (Page.IsPostBack)
            {
                string xml;
                if (Page.IsCallback)
                {
                    if (CallbackManager.CallbackID == null) CallbackManager.InitCallbackParams();
                    if (CallbackManager.CallbackID == UniqueID)
                    {
                        xml = CallbackManager.CallbackParameters.ControlReserved;
                        _Data = SerializableObjectBase.CreateFromXML<TextBoxData>(xml);
                        Text = _Data.Text;
                    }
                    else if (CallbackManager.CallbackParameters.CallbackOptions.PostDataContainer != "NA")
                    {
                        InitPostData();
                    }
                }
                else
                {
                    InitPostData();
                }
            }
        }

        protected override void OnBeforeRender()
        {
            _Settings.CallbackManagerClientInstance = CallbackManager.ClientInstanceName;
            _Settings.UniqueID = this.UniqueID;
            _Settings.EventTriggerHierarchyID = this.EventTriggerHierarchyID;
            _Settings.CommandArg = this.CommandArg;

            _Settings.ServerEvents_KeyDown = (KeyDown == null) ? false : true;
            _Settings.ServerEvents_KeyUp = (KeyUp == null) ? false : true;
            _Settings.ServerEvents_GetFocus = (GetFocus == null) ? false : true;
            _Settings.ServerEvents_LostFocus = (LostFocus == null) ? false : true;

            if (!ReadOnly && Enabled)
            {
                if (!CallbackManager.Scripts.IsScriptRegistered("TextBoxJs")) { CallbackManager.Scripts.RegisterClientScriptInclude("TextBoxJs", Page.ClientScript.GetWebResourceUrl(this.GetType(), "ZYH.WebControl_V.Js.TextBox.js")); }
                string st = "";
                st += "var " + ClientID + "Instance = new Class_TextBox('" + this.ClientID + "'," + base._CallbackOptions.ToJson_Js() + "," + _Settings.ToJson_Js() + ");\r\n";
                st += "" + ClientID + "Instance.Init();\r\n";
                CallbackManager.Scripts.RegisterStartupScript(ClientID + "InitJs", st);
            }
            base.OnBeforeRender();
        }

        protected override void RenderContents(System.Web.UI.HtmlTextWriter writer)
        {
            _TextBox.Attributes.Add("name", UniqueID); _TextBox.Attributes.Add("id", ClientID);
            _TextBox.ControlStyle.CopyFrom(this.ControlStyle);
            if (ReadOnly)
            {
                _TextBox.Attributes.Add("readonly", "readonly");
                _TextBox.Style.Add("background-color", System.Drawing.ColorTranslator.ToHtml(ReadOnlyBackColor));
            }
            writer.Write(RenderControl(_TextBox));
        }

        public override void RenderBeginTag(System.Web.UI.HtmlTextWriter writer) { }
        public override void RenderEndTag(System.Web.UI.HtmlTextWriter writer) { }
    }
}
